lib/types/src/schema.ts

```diff
@@ -1,92 +1,62 @@
 import type prompts from 'prompts'
 import type { Logger } from 'winston'
+import { z } from 'zod'
 
-export type ScalarSchemaType = 'string' | 'number' | 'boolean' | `|${string},${string}` | 'unknown'
-export type SchemaType = ScalarSchemaType | `array.${ScalarSchemaType}` | `record.${ScalarSchemaType}`
 export type SchemaPromptGenerator<T> = (
   logger: Logger,
   prompt: typeof prompts,
   onCancel: () => void
 ) => Promise<T>
-export type ModifiedSchemaType = SchemaType | `${SchemaType}?` | SchemaPromptGenerator<unknown>
-
-export type Schema = {
-  readonly [option: string]: ModifiedSchemaType
-}
-
-// Achieve the mapping with conditional types
-type SchemaTypeOutput<T extends SchemaType> = T extends 'string'
-  ? string
-  : T extends 'number'
-  ? number
-  : T extends 'boolean'
-  ? boolean
-  : T extends `|${infer A},${infer B}`
-  ? A | B
-  : T extends 'unknown'
-  ? unknown
-  : T extends `array.${infer S}`
-  ? S extends SchemaType
-    ? Array<SchemaTypeOutput<S>>
-    : never
-  : T extends `record.${infer S}`
-  ? S extends SchemaType
-    ? Record<string, SchemaTypeOutput<S>>
-    : never
+export type PromptGenerators<T> = T extends z.ZodObject<infer Shape>
```

Comment:
**question:** what is `Shape` here?

Reply:
`Shape` is the name of a type we're declaring using the `infer` keyword. It's a way to extract a type parameter from a generic type like `Array<string>`. You could use `Array<infer Item>` to get an `Item` type, though it could equally be called the `Value` type or the `Ivo` type. The `infer` keyword is useable in these conditional types (i.e., types that looks like `T extends Foo<infer Bar> ? U : V`) which you can read more about in the [TypeScript handbook](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types).

In this case `Shape` will be the first type parameter of the `z.ZodObject` type. This type parameter is an object whose values are the zod schemas of each value in the object's schema. In practical terms it's the type of the object you pass to `z.object` in a schema definition. So if you declared a schema like
```ts
const example: z.ZodObject = z.object({
  configPath: z.string(),
  id: z.number()
})
```
The inferred `Shape` type (i.e., `ZodObject`'s first type parameter) would be the type of
```ts
{
  configPath: z.string(),
  id: z.number()
}
```
so something like
```ts
{
  configPath: z.ZodString,
  id: z.ZodNumber
}
```
In `PromptGenerators` we then pass each of `Shape`'s values (the zod schemas for each option) to the `SchemaPromptGenerator` generic type.

Reply:
🙏 Thank you for the detailed explanation